Spigot-Server-Patches/0358-Custom-plugin-advancement-API.patch

```diff
@@ -0,0 +1,199 @@
+             this.methodProfiler.a();
+         }
+ 
++        if(org.bukkit.craftbukkit.advancement.CraftAdvancementManager.needsReload) server.reloadData(); // Paper - custom advancements API
+         this.methodProfiler.a("root");
+         this.D();
+         if (i - this.Z >= 5000000000L) {
+@@ -1663,6 +1664,7 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
+             this.getAdvancementData().reload();
+             this.aL().f();
+             this.getPlayerList().reload();
++            org.bukkit.craftbukkit.advancement.CraftAdvancementManager.needsReload = false;
+         } else {
+             this.postToMainThread(this::reload);
+         }
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+index 9fe555986..a1538075a 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+@@ -51,6 +51,7 @@ import org.bukkit.configuration.ConfigurationSection;
+ import org.bukkit.configuration.file.YamlConfiguration;
+ import org.bukkit.configuration.serialization.ConfigurationSerialization;
+ import org.bukkit.conversations.Conversable;
++import org.bukkit.craftbukkit.advancement.CraftAdvancementManager;
+ import org.bukkit.craftbukkit.boss.CraftBossBar;
+ import org.bukkit.craftbukkit.command.VanillaCommandWrapper;
+ import org.bukkit.craftbukkit.entity.CraftPlayer;
+@@ -170,6 +171,7 @@ public final class CraftServer implements Server {
+     private WarningState warningState = WarningState.DEFAULT;
+     private final BooleanWrapper online = new BooleanWrapper();
+     public CraftScoreboardManager scoreboardManager;
++    public CraftAdvancementManager advancementManager = new CraftAdvancementManager(); // Paper - custom advancement API
+     public boolean playerCommandState;
+     private boolean printSaveWarning;
+     private CraftIconCache icon;
+@@ -1738,6 +1740,12 @@ public final class CraftServer implements Server {
+         return scoreboardManager;
+     }
+ 
++    // Paper start - custom advancement API
++    public CraftAdvancementManager getAdvancementManager() {
++        return advancementManager;
++    }
++    // Paper end
++
+     public void checkSaveState() {
+         if (this.playerCommandState || this.printSaveWarning || this.console.autosavePeriod <= 0) {
+             return;
+diff --git a/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementManager.java b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementManager.java
+new file mode 100644
+index 000000000..a2603305e
+--- /dev/null
++++ b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementManager.java
+@@ -0,0 +1,96 @@
++package org.bukkit.craftbukkit.advancement;
++
++import com.google.common.collect.LinkedListMultimap;
++import com.google.common.collect.ListMultimap;
++import com.google.common.collect.Maps;
++import net.minecraft.server.*;
```

Comment:
http://www.javadude.com/articles/importondemandisevil.html

Reply:
IDEA's fault :frowning:

Reply:
Configure your IDE to not use * imports by setting the number of required imports to something *much* higher